File: 프로그래머스/3/42861. 섬 연결하기/섬 연결하기.js
function find(parent, x) {
    if (parent[x] !== x) {
        parent[x] = find(parent, parent[x]);
    }
    return parent[x];
}

function union(parent, a, b) {
    let pA = find(parent, a);
    let pB = find(parent, b);
    
    if (pA < pB) parent[pB] = pA;
    else parent[pA] = pB;   
}

function solution(n, costs) {
    costs.sort((a, b) => a[2] - b[2]);
    
    let parent = Array.from({length: n}, (_, i) => i);
    let total = 0;
    
    costs.forEach((cost) => {
        if (find(parent, cost[0]) !== find(parent, cost[1])) {
            union(parent, cost[0], cost[1]);
            total += cost[2];
        }
    })
    
    return total;
}

File: 프로그래머스/3/42861. 섬 연결하기/섬 연결하기.java
import java.util.*;

class Solution {
    static int[] parent;
    
    public int solution(int n, int[][] costs) {
        parent = new int[n + 1];
        for (int i = 1; i <= n; i++) {
            parent[i] = i;
        }
        
        Arrays.sort(costs, Comparator.comparingInt(o -> o[2]));
        
        int answer = 0;        
        for (int[] cost : costs) {
            if (find(cost[0]) != find(cost[1])) {
                union(cost[0], cost[1]);
                answer += cost[2];
            }
        }
        
        return answer;
    }
    
    static boolean union(int a, int b) {
        int pa = find(a);
        int pb = find(b);
        if (pa == pb) return false;
        
        if (pa < pb) parent[pb] = pa;
        else parent[pa] = pb;
        return true;
    }
    
    static int find(int x) {
        if (parent[x] != x) {
            parent[x] = find(parent[x]);
        }
        return parent[x];
    }
}